src/Container/Docker/Manifest.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Container.Docker.Manifest (
+  ManifestJson (..),
+  ManifestJsonImageEntry (..),
+  decodeManifestJson,
+  getLayerPaths,
+  manifestFilename,
+) where
+
+import Control.DeepSeq (NFData)
+import Data.Aeson (
+  FromJSON,
+  ToJSON,
+  defaultOptions,
+  eitherDecode,
+  genericToEncoding,
+  parseJSON,
+  toEncoding,
+  withObject,
+  (.!=),
+  (.:),
+ )
+import Data.ByteString.Lazy qualified as ByteStringLazy
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Text (Text)
+import GHC.Generics (Generic)
+
+manifestFilename :: Text
+manifestFilename = "manifest.json"
+
+-- | The manifest of container image for the top-level image, and optionally for parent images
+-- that this image was derived from. It consists of an array of metadata entries.
+--
+-- Reference:
+-- * https://github.com/moby/moby/blob/master/image/spec/v1.2.md
+-- * https://github.com/moby/moby/blob/master/image/spec/v1.1.md
+newtype ManifestJson = ManifestJson (NonEmpty.NonEmpty ManifestJsonImageEntry)
+  deriving (Show, Eq, Ord, Generic)
+  deriving anyclass (NFData) -- explicitly use DeriveAnyClass strategy to avoid deriving-defaults warning.
+
+instance FromJSON ManifestJson
+instance ToJSON ManifestJson where
+  toEncoding = genericToEncoding defaultOptions
+
+data ManifestJsonImageEntry = ManifestJsonImageEntry
+  { config :: Text -- references another file in the tar which includes the image JSON for this image.
+  , repoTags :: [Text] -- references pointing to this image.
+  , layers :: NonEmpty.NonEmpty FilePath -- points to the filesystem changeset tars
```

Comment:
Why does `layers` need to be a `NonEmpty FilePath`? Doesn't `FromJSON` already guarantee that it will parse a `FilePath` or it will fail?

Reply:
yes - but the container will be of type `[FilePath]`, which does not guarantee at least one layer property; Also, it's easier to perform `head` safely later on biz code.

Reply:
Makes sense, thanks for the clarification!